Fix: Display error for broken social connections in editor

gmjuhasz opened the pull request:
#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->

This change introduces a notice if there is any broken connection, and disables broken connections 
Added new property to connection meta to state health of connection. We already know this information, since this is cached for 12 hours, so we can easily use it in the requests. This also had an advantage, that in the editor the Notice is instantly visible, rather than before when it sent a request on page load to get the (same cached) connection results from the API, and then showing the notice.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

1202807135507995-as-1202841973973991

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Create a broken connection for facebook: 
  * Go [here](https://www.facebook.com/settings?tab=business_tools&ref=business_login_reentry) and revoke access for WordPress
  * In sandbox open wp-content/lib/external-connections.php file, and on line 655 change `$force = false` to `$force = true`
  * Open Social on your page, then go to Add new Post, and open the Social/Jetpack tab. This should refresh the test result and after that in cloud/Calypso the FB connection should show as broken.
  * Revert the change in the parameters, set it back to false. Now you have a broken connection without have to wait for 12 hours.
* Open Gutenberg, and go to Social tab. Facebook connection should be shown as broken and disabled:


Gutenberg            |  Classic
:-------------------------:|:-------------------------:
<img width="279" alt="CleanShot 2022-08-24 at 14 00 45@2x" src="https://user-images.githubusercontent.com/36671565/186413178-4b9b55ab-8015-49e7-82f4-257efe8e2c31.png"> |  <img width="290" alt="CleanShot 2022-08-29 at 14 27 16@2x" src="https://user-images.githubusercontent.com/36671565/187200853-3848aab9-d97a-4225-8f62-abf7eeb825dd.png">>


* Click on link, which navigates to JP cloud, and fix the connection.
* Going back to the Gutenberg tab, the connection should refresh in a few seconds. Notice should disappear, and connection can be toggled again. In Classic Editor you need to refresh the page to see changes
* Check this for both Social standalone and Jetpack version.

gmjuhasz committed:
[not verified] Rebase on parent commit

gmjuhasz committed:
[not verified] Rebase on parent commit

gmjuhasz committed:
[not verified] changelog

gmjuhasz committed:
[not verified] Add prop to refresh endpoint

gmjuhasz committed:
[not verified] Add to Classic Editor

gmjuhasz committed:
[not verified] Add to Classic Editor

gmjuhasz committed:
[not verified] Fix publicize package tests

gmjuhasz committed:
[not verified] changelog

spsiddarthan commented on projects/packages/publicize/src/class-publicize-ui.php:
```diff
@@ -373,7 +373,7 @@ public function post_page_metabox_assets() {
 		// If we have the #pub-connection-tests div present, kick off the connection test
 		if ( $( '#pub-connection-tests' ).length ) {
-			publicizeConnTestStart();
+			// publicizeConnTestStart(); TODO: Fix magic links
```
What's the significance of commenting this out and what are the magic links we are talking about?

gmjuhasz replied:
Sorry for not being clear on this! The connection tester function returns a link for each reparable connection which directly directs you to Facebook for example to fix the connection. These links are broken now. 
More: p1661505891955819-slack-C02JJ910CNL

I commented it out, so while it is not working it is not reaching out for the API introducing some extra calls, however as discussed if they would work it would be nice that the user could solve the issue right from the editor, so after looking up why are they broken and fixing them, we could reenable and refactor this.

spsiddarthan replied:
No issues, I appreciate the explanation, maybe we can add your explanation as the comment? :)

spsiddarthan commented on projects/packages/publicize/src/class-publicize-ui.php:
```diff
@@ -668,6 +672,27 @@ class="wpas-submit-<?php echo esc_attr( $connection_data['service_name'] ); ?>"
+					printf(
+						wp_kses(
+							/* translators: %s is the link to the connections page in Calypso */
+							__( 'Some of your social connections are broken! Reconnect them on the <a href="%s" rel="noopener noreferrer" target="_blank">connection management</a> page..', 'jetpack-publicize-pkg' ),
```
Confidence check:  Is `jetpack-publicize-pkg` a valid value for the second param here? I have seen only `jetpack` and `jetpack-social` thus far.

danielpost replied:
Yeah, it's the domain used for the Publicize PHP package 👍🏼

spsiddarthan commented on projects/plugins/jetpack/extensions/plugins/publicize/components/panel/index.js:
```diff
@@ -91,6 +94,7 @@ const PublicizePanel = ( { prePublish } ) => {
 						isPublicizeEnabled={ isPublicizeEnabled }
 						isRePublicizeFeatureEnabled={ isRePublicizeFeatureEnabled }
 						isPublicizeDisabledBySitePlan={ isPublicizeDisabledBySitePlan }
+						connectionsAdminUrl={ connectionsAdminUrl }
```
@pablinos @danielpost  Should we be taking the user may be to Jetpack Cloud here instead of https://wordpress.com/marketing/connections/<url> or is this ok to do?

pablinos replied:
Yes, I think we have a connections URL we can use in the Jetpack redirects. It should already be available as we use it for the ['Connect an account' link](https://github.com/Automattic/jetpack/blob/2447e267e40f2d133a5c15c2320b776e4c395ec8/projects/js-packages/publicize-components/src/components/settings-button/index.js#L36)

gmjuhasz replied:
I tried this, but the `getJetpackData()?.publicizeConnectionsUrl` is not available here 🤔

pablinos replied:
For Jetpack it isn't currently. [We default to the Calypso URL](https://github.com/Automattic/jetpack/blob/2447e267e40f2d133a5c15c2320b776e4c395ec8/projects/js-packages/publicize-components/src/components/settings-button/index.js#L36) when it's not available. We could maybe add that URL to `usePublicizeConfig`

gmjuhasz replied:
Good idea! Now I'll use the one we use with the ['Connect an account' link](https://github.com/Automattic/jetpack/blob/2447e267e40f2d133a5c15c2320b776e4c395ec8/projects/js-packages/publicize-components/src/components/settings-button/index.js#L36) then.

spsiddarthan reviewed (commented):
I got a strange error saying this class was not found `new Automattic\Jetpack\Social\Share_Limits` , I just uncommented that bit of code to continue to testing, but I wonder if this was something to do with my localhost not having the dependencies correctly installed. 

Things work well per the description, I have asked a question on the magic links and wonder if we should redirect to the Jetpack Cloud screen to fix the broken connections from social. Otherwise, it works perfectly! 


<img width="340" alt="Screenshot 2022-08-30 at 8 13 31 PM" src="https://user-images.githubusercontent.com/6594561/187470857-d9b22035-9e09-45c8-91ee-659412e855b0.png">

gmjuhasz commented:
> I got a strange error saying this class was not found `new Automattic\Jetpack\Social\Share_Limits` , I just uncommented that bit of code to continue to testing, but I wonder if this was something to do with my localhost not having the dependencies correctly installed.

Maybe `jetpack install --all` could solve it? I have had something similar where an install solved the issue

gmjuhasz committed:
[not verified] Fix tests

gmjuhasz committed:
[not verified] Revert last commit

gmjuhasz committed:
[not verified] Fix tests in plugins/jetpack

gmjuhasz committed:
[not verified] Rebase

pablinos commented on projects/plugins/social/src/js/components/panel/index.jsx:
```diff
@@ -29,12 +29,18 @@ const PublicizePanel = ( { prePublish } ) => {
+		numberOfSharesRemaining,
+		hasPaidPlan,
+		connectionsAdminUrl,
+	} = useSelect( select => {
```
In #25993 I've started refactoring this, so that we get this information from the editor initial state. I don't think we need to change anything here, but it's something to be aware of.

pablinos commented on projects/plugins/jetpack/_inc/lib/core-api/wpcom-fields/post-fields-publicize-connections.php:
```diff
@@ -145,6 +145,12 @@ private function post_connection_schema() {
 					'context'     => array( 'edit' ),
 					'readonly'    => true,
 				),
+				'is_healthy'      => array(
```
I don't think we want to use this version of the file anymore.

gmjuhasz replied:
Oh, okay, removed it!

pablinos reviewed (commented):
Just a couple of thoughts around the connection URL. I'll give it a proper test shortly.

gmjuhasz committed:
[not verified] Address comments

gmjuhasz committed:
[not verified] Changelogger validity

gmjuhasz committed:
[not verified] Remove file

gmjuhasz committed:
[not verified] Re-add prop to post fields

spsiddarthan commented:
This is working fine, @gmjuhasz! Thanks for incorporating the comments. I can give another +1 after the tests are updated!

gmjuhasz committed:
[not verified] Change mocking to conditional

spsiddarthan reviewed (dismissed):
This is looking good!

gmjuhasz committed:
Move files to other PR

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

pablinos committed:
Social: Change the connections endpoint to connection-test-results

After merging #26216, we discovered that Jetpack Social was not using
the same API endpoint to get the connection data in the editor. The one
used by Jetpack and WPCOM already provided information about the health
of the connection.

This change reverts the addition of the `is_healthy` property in the
`/publicize/connections` endpoint and updates Jetpack Social to use the
`/publicize/connection-test-results` endpoint.

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Merge remote-tracking branch 'origin/update/social-connections-endpoint' into add/error-for-broken-connection

gmjuhasz committed:
Change is_healthy check to test_success in effects

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Connection url followup change

gmjuhasz committed:
Update package dependencies

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Fixup package versions

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Update package.json

pablinos commented on projects/plugins/jetpack/class.jetpack-gutenberg.php:
```diff
@@ -711,8 +711,11 @@ public static function enqueue_block_editor_assets() {
+				'sharesData'              => $publicize->get_publicize_shares_info( $blog_id ),
+				'hasPaidPlan'             => $publicize->has_paid_plan(),
+				'publicizeConnectionsUrl' => esc_url_raw(
+					'https://jetpack.com/redirect/?source=jetpack-social-connections-block-editor&site='
```
I wonder if we should leave this as having the default of https://wordpress.com/marketing/connections because this will be used on WPCOM too, and it seems odd to send them via a Jetpack redirect.

Moving it to the usePublicizeConfig hook seems useful though, so it can be used in various places, rather than just the share panel.

It's also taking me to wordpress.com rather than cloud.jetpack.com when I'm using Jetpack Social

gmjuhasz replied:
I removed this, and changed the logic of the connectionURL so it will be the same as the one for the "Connect an account" button below. I moved that to the usePublicizeConfig hook.

pablinos commented on projects/packages/publicize/src/class-publicize-ui.php:
```diff
@@ -691,6 +695,27 @@ class="wpas-submit-<?php echo esc_attr( $connection_data['service_name'] ); ?>"
+					printf(
+						wp_kses(
+							/* translators: %s is the link to the connections page in Calypso */
+							__( 'Some of your social connections are broken! Reconnect them on the <a href="%s" rel="noopener noreferrer" target="_blank">connection management</a> page..', 'jetpack-publicize-pkg' ),
```
Do you think we could get a count of the unhealthy connections? That way we could have another variation of this that says "One of your social..."

I wonder if we should not use an exclamation mark too. It should be a bit urgent, but we don't want to painc them

gmjuhasz replied:
Good idea I change this!

pablinos commented on projects/packages/publicize/src/class-publicize-ui.php:
```diff
@@ -374,7 +374,7 @@ public function post_page_metabox_assets() {
 		// If we have the #pub-connection-tests div present, kick off the connection test
 		if ( $( '#pub-connection-tests' ).length ) {
-			publicizeConnTestStart();
+			// publicizeConnTestStart(); TODO: Fix magic links
```
What's involved with fixing this? Is it something we should do now?

gmjuhasz replied:
Well, probably not. I think it needs more investigation beforehand, something is up with the logic of that. I would rather leave this as is for now and fix this with the design update together in the next task. What do you think?

pablinos replied:
Fair enough. Do you think we should just get rid of this code block for now? It seems to only exist to call this function, which isn't happening currently.

gmjuhasz replied:
Yeah makes sense!

pablinos reviewed (commented):
I think this is nearly there. I just had some thoughts about the connection URL and the notice that's displayed. It's working well though.

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Addressing review comments

gmjuhasz committed:
Fix composer.lock

pablinos commented on projects/js-packages/publicize-components/src/hooks/use-publicize-config/index.js:
```diff
@@ -23,6 +24,8 @@ export default function usePublicizeConfig() {
 	);
 	const isPostPublished = useSelect( select => select( editorStore ).isCurrentPostPublished(), [] );
+	const connectionsRootUrl =
+		getJetpackData()?.publicizeConnectionsUrl ?? 'https://wordpress.com/marketing/connections/';
```
I think we still have [this nested in the `social` object](https://github.com/Automattic/jetpack/blob/5d9ce7c108c1c710d1d383fe3ea820fc613132b1/projects/plugins/social/src/class-jetpack-social.php#L257), so we should add that to the path here.

gmjuhasz replied:
Done

pablinos commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -82,16 +86,40 @@ export default function PublicizeForm( {
+						<Notice type={ 'error' }>
+							{ createInterpolateElement(
+								_n(
+									'One of your social connections is broken! Reconnect them on the <fixLink>connection management</fixLink> page.',
```
It would be good to remove the exclamation mark here too, to keep things consistent.

gmjuhasz replied:
Oh shoot, forgot here, done!

pablinos reviewed (commented):
This is working well for me. Thanks for making the changes. I think there's just a couple more bits and then we can get it merged.

gmjuhasz committed:
Fix connection url

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Update Composer.lock

pablinos commented on projects/js-packages/publicize-components/src/components/settings-button/index.js:
```diff
@@ -6,7 +6,7 @@
  * then connections will be automatically refreshed.
  */
-import { getSiteFragment, getJetpackData } from '@automattic/jetpack-shared-extension-utils';
+import { usePublicizeConfig } from '@automattic/jetpack-publicize-components';
```
I think `usePublicizeConfig` is local to this component now, so we could import it as `../../hooks/use-publicize-config`

gmjuhasz replied:
Done!

pablinos reviewed (dismissed):
This is working well for me. Just one super minor comment about the `usePublicizeConfig` import. I don't know if there is an advantage to doing it that way or not.

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

gmjuhasz committed:
Fix import and composer lock

pablinos commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -47,6 +49,8 @@ export default function PublicizeForm( {
 		! isRePublicizeFeatureEnabled && connections.every( connection => ! connection.toggleable );
 	const Wrapper = isPublicizeDisabledBySitePlan ? Disabled : Fragment;
 
+	const brokenConnections = connections.filter( connection => ! connection.is_healthy );
```
I think if we remove `is_healthy` from the post field data, then we might have to default it to `true` or do a strict check, like `false === connection.is_healthy` to make sure that it doesn't look like all the connections are broken before we get the response from the test connections API call.

pablinos reviewed (commented):
We might need to make some changes to this if we remove the `is_healthy` property from the post meta field. We might want to assume we are going to do that and make it so the `is_healthy` checks will still work.

pablinos committed:
[not verified] Refactor the classic editor connection test JS code.

pablinos committed:
[not verified] Correct usePublicizeConfig import.

pablinos committed:
[not verified] Correct the text domain.

pablinos committed:
[not verified] Make is_healthy checks strict.

pablinos committed:
[not verified] Remove is the connection test result from the post field.

pablinos commented:
@gmjuhasz I've done a bit of work on this, which turned out to be more than I imagined! Related to the fix in #26955 I decided we should remove `is_healthy` from the post field, which means we need to wait for the result of the initial API/AJAX request in the editor. I don't think that's a major issue, given the performance problem having the connection tests creates.

But, this led to a pretty major refactor of the classic editor code, so that I could use `_n` for i18n of the strings etc. It should refresh on window foces though, like we do in the block editor, so it's a bit closer to the same UX.

See what you think, and if you're happy we can get others on the team and Crew to take a look.

kraftbj commented:
Based on this comment in Slack p1666385236922229/1664903339.698629-slack-CDLH4C1UZ , I am marking this for 11.5 milestone. This will keep it on the Releasers radar to see where y'all are at if it is still open when they're looking at cutting.

gmjuhasz committed:
Merge branch 'trunk' into add/error-for-broken-connection

pablinos committed:
Merge remote-tracking branch 'origin/trunk' into add/error-for-broken-connection

danielpost commented on projects/js-packages/publicize-components/src/components/form/index.js:
```diff
@@ -82,16 +86,40 @@ export default function PublicizeForm( {
+											false === is_healthy
 										}
-										enabled={ enabled && ! isPublicizeDisabledBySitePlan }
+										enabled={ enabled && ! isPublicizeDisabledBySitePlan && false !== is_healthy }
```
Nitpick, but this reads a little awkward. Isn't `false !== is_healthy` the same as just `is_healthy`?

danielpost replied:
Reading back this might actually be intended, so feel free to ignore!

pablinos replied:
Yeah we need to check it's actually `false`

pablinos committed:
Merge remote-tracking branch 'origin/trunk' into add/error-for-broken-connection
